pipeline/load/display_metrics.py

```diff
@@ -1,59 +1,34 @@
-import operator
+from typing import Dict, List
 
 import prefect
 from prefect import task
 
-from pipeline.types import MatchingMetrics, MatchingOutput, User
+from pipeline.types import MatchingMetrics, MatchingOutput, User, UserId
+
+
+def render_counts_per_user(
+    users: List[User], counts: Dict[UserId, int]
+) -> List[str]:
+    """Helper method to get lines of counts per user in descending order."""
+    display_names = {user.uid: user.displayName for user in users}
+    # Sort entries by count in descending order
+    sorted_items = sorted(counts.items(), key=lambda p: p[1], reverse=True)
+    lines = []
+    for uid, count in sorted_items:
+        name = display_names.get(uid)
+        lines.append(f"{count} - {name} ({uid})")
+    return lines
 
 
 @task
 def display_internal_matching_metrics(
-    output: MatchingOutput, metric: MatchingMetrics, return_metric=False
+    output: MatchingOutput, metrics: MatchingMetrics
 ):
+    """Task to display matching engine metrics."""
     logger = prefect.context.get("logger")
-    display_names = [user.displayName for user in output.users]
-    unique_names = set()
-    duplicate_names = []
-    for name in display_names:
-        if name in unique_names:
-            duplicate_names.append(name)
-        else:
-            unique_names.add(name)
-    unique_users = [
-        user for user in output.users if user.displayName not in duplicate_names
-    ]
-    alter_metric_output = {
-        uid.replace(
-            uid,
-            next(
-                (user.displayName for user in unique_users if user.uid == uid),
-                "Unknown",
-            ),
-        ): count
-        for uid, count in metric.n_proposed_matches_per_user.items()
-    }
-    if "Unknown" in alter_metric_output:
-        del alter_metric_output["Unknown"]
-    for user in output.users:
-        if user.displayName in duplicate_names:
-            alter_metric_output[
-                f"{user.displayName} (uid: {user.uid})"
-            ] = metric.n_proposed_matches_per_user[user.uid]
-    n_proposed_matches_per_user = dict(
-        sorted(
-            alter_metric_output.items(),
-            key=operator.itemgetter(1),
-            reverse=True,
-        )
-    )
-    if return_metric:
-        return n_proposed_matches_per_user
-    logger.info("Generating internal matching metrics")
-    logger.info("Displaying number of proposed matches per user")
-    logger.info(
-        "Matches Proposed per User:\n"
-        + "\n".join(
-            f"{count} - {user}"
-            for user, count in n_proposed_matches_per_user.items()
-        )
+    lines = render_counts_per_user(
+        output.users, metrics.n_proposed_matches_per_user
     )
+    logger.info("Matches Proposed per User:")
+    for line in lines:
+        logger.info(line)
```

Comment:
Now the display task is very simple: it just logs the lines returned by the helper function.

Reply:
This is a great point! I agree that it would be better to use just one log per metric rather than one log per line.

I revised the display method to return a multi-line string rather than a list of strings, updated the test, and then updated the logging code.